### PR TITLE
raftstore: region initial size depends on the split resource .

### LINKED ISSUE #15457: region size maybe not accurate after region split
## Development Task
In past, the new region size maybe the half of the old region after the region splitted. But sometime, the new region's key range don't contains any data suck like `create new table`. 
The new region's size should depends on the split policy. 
- region size should be zero if the policy is usekey by the user.
- region size should be the half size if the policy is load split like size check or load split.

### PR DESCRIPTION
**Pull request by bufferflies**
### What is changed and how it works?

Issue Number: Close #15457
depends: https://github.com/pingcap/kvproto/pull/1177
What's Changed:

```commit-message
there are three triggers will split the regions:
1. load split include sizekeys, load etc. In this cases, the new region should contains the data  after split.
2. tidb split tables or partition table, such like `create table test.t1(id int,b int) shard_row_id_bits=4 partition by hash(id) partitions 2000`. 
In this cases , the new region shouldn't contains any data  after split.

```

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)

image
image

Side effects

### Release note 

```release-note
None
```

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -201,8 +201,8 @@ procinfo = { git = "https://github.com/tikv/procinfo-rs", rev = "6599eb9dca74229
 # When you modify TiKV cooperatively with kvproto, this will be useful to submit the PR to TiKV and the PR to
 # kvproto at the same time.
 # After the PR to kvproto is merged, remember to comment this out and run `cargo update -p kvproto`.
-# [patch.'https://github.com/pingcap/kvproto']
-# kvproto = { git = "https://github.com/your_github_id/kvproto", branch = "your_branch" }
+[patch.'https://github.com/pingcap/kvproto']
+kvproto = { git = "https://github.com/bufferflies/kvproto", branch = "split_size" }
```
Comment (overvenus) [inline Cargo.toml:205]: Need to update kvproto first.

File: components/raftstore/src/store/worker/pd.rs [hunk 2]
```diff
@@ -2381,24 +2394,28 @@ fn new_split_region_request(
...
 fn new_batch_split_region_request(
     split_keys: Vec<Vec<u8>>,
     ids: Vec<pdpb::SplitId>,
     right_derive: bool,
+    skip_size_check: bool,
 ) -> AdminRequest {
     let mut req = AdminRequest::default();
     req.set_cmd_type(AdminCmdType::BatchSplit);
     req.mut_splits().set_right_derive(right_derive);
+    req.mut_splits().set_skip_size_check(skip_size_check);
```
Comment (overvenus) [inline components/raftstore/src/store/worker/pd.rs:2418]: What does  `skip_size_check`?

Comment (bufferflies) [reply components/raftstore/src/store/worker/pd.rs:2418]: if `skip_size_check` is true, the new splitted region's initial size is zero not the half size of the source region.

Comment (overvenus) [reply components/raftstore/src/store/worker/pd.rs:2418]: How about `inherit_size`?

Comment (bufferflies) [reply components/raftstore/src/store/worker/pd.rs]: has change as amortize_source_region_size.

File: components/raftstore-v2/src/operation/command/admin/split.rs [hunk 3]
```diff
@@ -148,6 +149,7 @@ pub struct RequestSplit {
     pub epoch: RegionEpoch,
     pub split_keys: Vec<Vec<u8>>,
     pub source: Cow<'static, str>,
+    pub amortize_source_region_size: bool,
```
Comment (SpadeA-Tang) [inline components/raftstore-v2/src/operation/command/admin/split.rs:152]: Please add some comments to describe it.

Comment (SpadeA-Tang) [reply components/raftstore-v2/src/operation/command/admin/split.rs:152]: Including which cases it will be true/false.

**Review by SpadeA-Tang — APPROVED** [review]

File: Cargo.toml (hunk 1 above)
Comment (bufferflies) [reply Cargo.toml]: fixed

File: components/raftstore-v2/src/operation/command/admin/split.rs (hunk 3 above)
Comment (bufferflies) [reply components/raftstore-v2/src/operation/command/admin/split.rs:156]: fixed

File: components/raftstore-v2/src/operation/command/admin/split.rs [hunk 4]
```diff
@@ -75,6 +75,9 @@ pub struct SplitResult {
     // The index of the derived region in `regions`
     pub derived_index: usize,
     pub tablet_index: u64,
+    // new regions will amortize the region size if it's true.
+    // otherwise, the new region's size will be 0.
+    pub amortize_source_region_size: bool,
```
Comment (Connor1996) [inline components/raftstore-v2/src/operation/command/admin/split.rs:80]: `amortize` means gradually write off the initial cost of (an asset) over a period. It's not proper here.

Comment (tonyxuqqi) [reply components/raftstore-v2/src/operation/command/admin/split.rs:80]: +1

Comment (tonyxuqqi) [reply components/raftstore-v2/src/operation/command/admin/split.rs:80]: I think the name can be size_based_split: bool

Comment (bufferflies) [reply components/raftstore-v2/src/operation/command/admin/split.rs:80]: has renamed `share_source_region_size `.

File: components/raftstore-v2/src/operation/command/admin/split.rs [hunk 5]
```diff
@@ -695,8 +705,14 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
 
         let new_region_count = res.regions.len() as u64;
         let control = self.split_flow_control_mut();
-        let estimated_size = control.approximate_size.map(|v| v / new_region_count);
-        let estimated_keys = control.approximate_keys.map(|v| v / new_region_count);
+        // if share_source_region_size is true, it means the new region contains any
+        // data from the origin region.
+        let mut amortize_size = None;
```
Comment (tonyxuqqi) [inline components/raftstore-v2/src/operation/command/admin/split.rs:710]: rename these amortize_xxx as well

**Review by tonyxuqqi — APPROVED** [review]

**Pull request merged**